Version2 development preprocessor nomenclature 847

ledm opened the pull request:
This PR addresses #847 (partially). 

In this patch we:

* rename `_area_pp.py`  to `_area.py`
* rename `_volume_pp.py`  to `_volume.py`
* rename `_time_area.py`  to `_time.py`

It also renames some preprocessors that were previously imported under a different name. This should ensure that the recipes match the API. 

As a bonus, I've added some basic documentation for these three preprocessor modules.

valeriupredoi commented on esmvaltool/preprocessor/_volume.py:
```diff
@@ -14,9 +14,9 @@
+def extract_volume(cube, z_min, z_max):
     """
-    Subset a cube on volume
+    Subset a cube based on a depth range.
```
could work for air pressure range as well no?

ledm commented:
Cheers @valeriupredoi. It wasn't so bad, just a few find replace commands and a fairly heavy reliance on `grep -ir`. I feel some responsibility too as I wrote several of these preprocessors in the first place.

ledm commented:
I've replaced several instances of the term `depth` in the documentation where appropriate. 

Although this goes beyond the scope of this PR, your comment uncovered one thing I don't fully understand here. That's how this part of the function `extract_volume` works:

```
    subz = iris.Constraint(
        depth=lambda cell: float(zmin) <= cell <= float(zmax))
```
I don't get how iris knows to apply this to the depth dimension. `depth` is only used to name the lambda function, and python doesn't really care about object names. How does iris know to use this on the depth?

The other problem here is if iris is doing something clever using the name of the lambda function, then this can only apply to the ocean. We'd need to write a second constraint for the atmosphere. @valeriupredoi , can you help me make this work for other z direction fields (like air pressure, or whatever they use in the atmosphere)? I've changed it to:
```
    cube_dims = [c.standard_name for c in cube.coords()]
    if 'depth' in cube_dims:
        z_constraint = iris.Constraint(
            depth=lambda cell: float(zmin) <= cell <= float(zmax))
```
Once again, this is beyond the scope of this PR, but seems like as good a place as any to do this.

valeriupredoi commented:
> I've replaced several instances of the term `depth` in the documentation where appropriate.
> 
> Although this goes beyond the scope of this PR, your comment uncovered one thing I don't fully understand here. That's how this part of the function `extract_volume` works:
> 
> ```
>     subz = iris.Constraint(
>         depth=lambda cell: float(zmin) <= cell <= float(zmax))
> ```
> 
> I don't get how iris knows to apply this to the depth dimension. `depth` is only used to name the lambda function, and python doesn't really care about object names. How does iris know to use this on the depth?
> 

the `iris.Constraint` function uses the lambda function argument and matches it to the coordinate name that needs to apply the constraint on, see: https://scitools.org.uk/iris/docs/latest/iris/iris.html#iris.Constraint

> The other problem here is if iris is doing something clever using the name of the lambda function, then this can only apply to the ocean. We'd need to write a second constraint for the atmosphere. @valeriupredoi , can you help me make this work for other z direction fields (like air pressure, or whatever they use in the atmosphere)? I've changed it to:
> 
> ```
>     cube_dims = [c.standard_name for c in cube.coords()]
>     if 'depth' in cube_dims:
>         z_constraint = iris.Constraint(
>             depth=lambda cell: float(zmin) <= cell <= float(zmax))
> ```
> 
> Once again, this is beyond the scope of this PR, but seems like as good a place as any to do this.

very good point! I failed at noticing the ocean use only. It's easy to make it applicable to other z-coords by making the caller (lambda) func argument generic ie `zcoords = list_of_all_possible_zcoords` and then giving a specific constraint for each of those like in the example page above:
```
iris.Constraint(name='air_potential_temperature',
           cube_func=lambda cube: cube.units == 'kelvin',
           coord_values={'latitude':lambda cell: 0 < cell < 90},
           model_level_number=[10, 12])
           & Constraint(ensemble_member=2)
```

valeriupredoi commented:
but let's do that in a new PR, shall we? :taxi:

ledm commented:
Sure, I'll make and issue. This one is probably ready to go then!

valeriupredoi commented on tests/unit/preprocessor/_volume/test_volume.py:
```diff
@@ -44,7 +44,8 @@ def setUp(self):
             [0.5, 5., 50.],
             standard_name='depth',
```
you may want to add a separate test for `air_pressure` as well, or even better, change depth with a generic Z-axis coordinate `zcoord` and guess it (so that the attr axis gets added to the coordinate) like this:
```
self.cube.coord('zcoord').attributes['positive'] = 'up'
iris.util.guess_coord_axis(self.cube.coord('zcoord'))
```

ledm replied:
Hmmm.....
```
           ValueError: 'zcoord' is not a valid standard_name
```

valeriupredoi replied:
stay put, am fixing it straight in the PR now :radio:

valeriupredoi replied:
ok done, the thing is - `extract_volume` now works with __any__ generic z-coord without naming it explicitly (which is very good) but all the other functions need an explicit name of a zcoordinate - do you need to keep that behaviour or we can nuke all the coordz arguments and replace them with axis='z' in the function body?

valeriupredoi replied:
ah sod it, too much work :grin:

valeriupredoi replied:
actually I've done it :grin: now all those functions will work with any z coord without explicitly naming it - did I just kill all your oceans recipes?

ledm replied:
yes...

valeriupredoi replied:
feel free to revert, but I think we're in a better position now, both from a generality point of view and from user perspective (ie they don't have to dig in their data to figure out what the name of the vertical axis is)

valeriupredoi replied:
I can fix all the func args in the diags if you want me to :smiling_imp:

ledm replied:
Also all the documentation will need to be re-written.

valeriupredoi replied:
I did change the docstrings though :grin:

ledm replied:
Here's a question for you. If we can automatically figure out the name of the z direction, could we also do that with latitude and longitude? 

Do we need to  have coord1 and coord2 in (for instance):
```
    average_region:
      coord1: longitude
      coord2: latitude
```
?

valeriupredoi replied:
yes, in fact, we can get rid of all the explicit calls to latitude and longitude axes as well, in the same manner, wherever we collapse on all the values; we should use them only when we apply specific boundaries or individual operations on individual axes

valeriupredoi replied:
but we need to think carefully if we want to support anything that's not CMOR standard 9which we don't) since these axes = X, Y, Z and T are present only in CMORized files

ledm replied:
That must surely be for a future issue/PR though. Changing coordz was already well beyond the scope of this PR!

valeriupredoi replied:
yup, absolutely!

ledm commented:
So, we're good to go here?

valeriupredoi commented:
@bouweandela and @mattiarighi you guys green light this?

valeriupredoi commented:
mergy-merge?

ledm commented:
Any comments from @bouweandela before we merge?

ledm commented:
Can we merge this please?

valeriupredoi commented:
just one second, I will merge it as soon as we fix an accidental push to `version2_development` that happened yesterday; I'd keep all the mergers on standby until that gets fixed first @mattiarighi @bjoernbroetz have you guys fixed it yet? Ill have a look now....
